File: SourceAFIS/Extraction/Filters/ThresholdBinarizer.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using SourceAFIS.General;

namespace SourceAFIS.Extraction.Filters
{
    public sealed class ThresholdBinarizer
    {
        public BinaryMap Binarize(float[,] input, float[,] baseline, BinaryMap mask, BlockMap blocks)
        {
            BinaryMap binarized = new BinaryMap(input.GetLength(1), input.GetLength(0));
            for (int blockY = 0; blockY < blocks.AllBlocks.Height; ++blockY)
            {
                for (int blockX = 0; blockX < blocks.AllBlocks.Width; ++blockX)
                {
                    if (mask.GetBit(blockX, blockY))
                    {
                        RectangleC rect = blocks.BlockAreas[blockY, blockX];
                        for (int y = rect.Bottom; y < rect.Top; ++y)
                            for (int x = rect.Left; x < rect.Right; ++x)
                                if (input[y, x] - baseline[y, x] > 0)
                                    binarized.SetBitOne(x, y);
                    }
                }
            }
            return binarized;
        }
    }
}

File: java/sourceafis/src/sourceafis/extraction/filters/ThresholdBinarizer.java
/**
 * @author Veaceslav Dubenco
 * @since 18.10.2012
 */
package sourceafis.extraction.filters;

import sourceafis.general.BinaryMap;
import sourceafis.general.BlockMap;
import sourceafis.general.DetailLogger;
import sourceafis.general.RectangleC;
import sourceafis.general.parallel.Parallel;
import sourceafis.general.parallel.ParallelForDelegate;

/**
 * 
 */
public final class ThresholdBinarizer {
	public DetailLogger.Hook Logger = DetailLogger.off;

	public BinaryMap Binarize(final float[][] input, final float[][] baseline,
			final BinaryMap mask, final BlockMap blocks) {
		final BinaryMap binarized = new BinaryMap(input[0].length, input.length);

		ParallelForDelegate<Object> delegate = new ParallelForDelegate<Object>() {
			@Override
			public Object delegate(int blockY, Object obj) {
				for (int blockX = 0; blockX < blocks.getAllBlocks().Width; ++blockX) {
					if (mask.GetBit(blockX, blockY)) {
						RectangleC rect = blocks.getBlockAreas().get(blockY,
								blockX);
						for (int y = rect.getBottom(); y < rect.getTop(); ++y)
							for (int x = rect.getLeft(); x < rect.getRight(); ++x)
								if (input[y][x] - baseline[y][x] > 0)
									binarized.SetBitOne(x, y);
					}
				}
				return null;
			}

			@Override
			public Object combineResults(Object res1, Object res2) {
				return null;
			}
		};

		Parallel.For(0, blocks.getAllBlocks().Height, delegate, null);
		Logger.log(binarized);
		return binarized;
	}
}
